### PR TITLE
[FIX] Addresses prolonged freezes in some chat rooms

### PR DESCRIPTION
**Pull request by Sameesunkaria**
@RocketChat/ios

- The regex for emoji replacement is now fixed and does not go into a state of catastrophic backtracking.
- An edge case fixed where the emojis would not get replaced.
- Also, addresses the concerns raised by @cardoso in #1451 
- Added test cases to cover a variety of edge cases.

### CODE REVIEW

File: Rocket.Chat/External/RCEmojiKit/Emojione+Transform.swift [hunk 1]
```diff
@@ -9,7 +9,7 @@
 import Foundation
 
 extension Emojione {
-    static func transform(string: String) -> String {
+    static func transrform(string: String) -> String {
```
Comment (rafaelks) [inline Rocket.Chat/External/RCEmojiKit/Emojione+Transform.swift]: Ups, typo?

Comment (Sameesunkaria) [reply Rocket.Chat/External/RCEmojiKit/Emojione+Transform.swift]: Oh I meant to remove that function

File: Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift [hunk 2]
```diff
@@ -10,51 +10,65 @@ import Foundation
...
+        let matchRanges = ranges.map { range in regex?.matches(in: string.string, options: [], range: range).map { $0.range(at: 0) } ?? [] }
+        return matchRanges.reduce(into: [NSRange]()) { $0.append(contentsOf: $1) }
     }
 }
 
 extension String {
     func codeRanges() -> [NSRange] {
         let codeRegex = try? NSRegularExpression(pattern: "(```)(?:[a-zA-Z]+)?((?:.|\r|\n)*?)(```)", options: [.anchorsMatchLines])
-        let codeMatches = codeRegex?.matches(in: self, options: [], range: NSRange(location: 0, length: count)) ?? []
+        let codeMatches = codeRegex?.matches(in: self, options: [], range: NSRange(location: 0, length: utf16.count)) ?? []
```
Comment (rafaelks) [inline Rocket.Chat/External/RCEmojiKit/NSAttributedString+CustomEmojis.swift]: 😅
